apple/swift-docc-render
# Fix issues with losing state when navigating or opening/closing items, while filtering.

Bug/issue #, if applicable: 90843762

## Summary

This PR fixes a few issues we have noticed during testing:

- [x] rdar://90841554 ([TESTFEST] Navigation collapses on searched term when performing a search, expanding and then selecting a subsection)
- [x] rdar://90843762 ([TESTFEST] Filtering for Sample Code yields inconsistent UI experience)
- [x] Navigating across pages while filtering, re-opens all previously closed items.

## Dependencies

NA

## Testing

You can use any doccarchive with an index.

Steps:
1. Apply a generic enough filter
2. Start opening/closing items.
3. Assert there are no items being lost when opening/closing.
4. Try to find a parent item that has no children matches. Assert it renders as closed initially, but you can expand and see it's children.
5. Try cmd + click to open all siblings. Assert its all working as expected.
6. Try alt + click to open all nested children. Assert they are all rendered in correct order.
7. Close some items and start navigating across links in the page OR by clicking navigator items. 
9. Assert the collapsed items are not opened, unless your current page is in one of them.
10. Try the same with going back/forward in the browser.
11. Clear the filters and assert the only open items are the ones leading to the current page.
12. Assert that opening/closing items without a filter works as expected.

## Checklist

Make sure you check off the following items. If they cannot be completed, provide a reason.
